### DISCUSSION TITLE
[Question][Hw2] Questions on Skeleton code

### DISCUSSION THREAD

**Discussion by Lanph3re** [prosyslab-classroom/cs524-program-analysis#412] [Q&A]
Hello. I skimmed given code a bit and noticed the code lacks something.

Do we have TODO parts that aren't marked as `failwith "Not implemented`?

For example `let rec execute_block` is declared as a recursive function but it doesn't call itself.
Overall code doesn't have code that fetches next instruction or destination instruction of jmp command.
Probably I think that I need to implement the fetching code in `transfer` function, but the return type of the function is `Memory` which doesn't have program counter things.

Am I missing something? Thank you in advance.

**Comment by Maokami** [comment]
I think this link will help you :
[link](https://stackoverflow.com/questions/38884482/what-does-the-and-keyword-mean-in-ocaml/38884590)

The functions `execute_block`, `execute_instr` and `transfer` are mutually recursively defined.

**Comment by Lanph3re** [comment]
Oh, the link really helps, thank you!

**Comment by hyunsukimsokcho** [comment]
There is nothing wrong in the skeleton. Please try to take full advantage of given skeleton code. I assume you are frustrated since you have bare understandings about LLVM APIs in OCaml. We really recommend you to first thoroughly and carefully read all the contents in README (not just for HW2, but for those will follow through the semester). Although we totally understand that one is not familiar with LLVM API in OCaml at first, rather than nitpicking on the skeleton, please get your hands dirty first and come up with more specific issues so that we could directly help.

**Comment by Lanph3re** [comment]
Ok, I'll keep that in mind. Thank you.
